components/camera/replaypcd/replaypcd.go

```diff
@@ -135,38 +157,112 @@ func (replay *pcdCamera) NextPointCloud(ctx context.Context) (pointcloud.PointCl
 			SortOrder: datapb.Order_ORDER_ASCENDING,
 		},
 		CountOnly:     false,
-		IncludeBinary: true,
+		IncludeBinary: replay.limit == 1,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	// If no data is returned, return an error indicating we've reached the end of the dataset.
-	if len(resp.GetData()) == 0 {
-		return nil, errEndOfDataset
-	}
-
 	replay.lastData = resp.GetLast()
-	data := resp.Data[0].GetBinary()
-
-	r, err := gzip.NewReader(bytes.NewBuffer(data))
-	if err != nil {
-		return nil, errors.Errorf("Failed to initialize gzip reader: %v", err)
-	}
 
-	defer func() {
-		if err = r.Close(); err != nil {
-			replay.logger.Warnw("Failed to close gzip reader", "warn", err)
+	// If batching use metadata to download data in parallel, caching results
+	if replay.limit > 1 {
+		for _, dataResponse := range resp.Data {
+			newCacheEntry := cacheEntry{id: dataResponse.GetMetadata().Id}
+			replay.cache = append(replay.cache, newCacheEntry)
 		}
-	}()
+		ctxTimeout, cancelTimeout := context.WithTimeout(ctx, downloadTimeout)
+		defer cancelTimeout()
 
-	pc, err := pointcloud.ReadPCD(r)
+		replay.downloadBatch(ctxTimeout)
+		replay.waitForDataAndUpdateCache(ctxTimeout)
+		return replay.NextPointCloud(ctx)
+	}
+
+	// If not batching, decode the returned binary directly
+	pc, err := decodeResponseData(resp.GetData(), replay.logger)
 	if err != nil {
 		return nil, err
 	}
+
 	return pc, nil
 }
 
+// downloadBatch iterates through the current cache, performing the download of the respective data in
+// parallel using go routines.
+func (replay *pcdCamera) downloadBatch(ctx context.Context) {
+	var wg sync.WaitGroup
+
+	// Parallelize download of data based on given ids
+	for _, cEntry := range replay.cache {
+		wg.Add(1)
+		go func(id string) {
```

Comment:
This should be a PanicCapturingGo. If it isn't an a panic happens, the whole process will crash which can disrupt other robot functionality